The third generation sequencing can generate multi-kilobase sequences and has the potential to improve genome assembly. Nevertheless, it has higher error rates in comparison with second generation sequencing. The error rates have limited its use to improve assembly. In this thesis, we introduce a hybrid correction algorithm to correct third-generation sequencing reads by finding overlapping short reads with high-quality. We improved the efficiency of a previous method which align short reads onto long PacBio reads using FM-index. The results indicate that the accuracy of corrected PacBio reads achieve over 93\%, the memory consumption is lower, and the running time is faster than previous method.

Keywords: Alignment, Second generation sequencing, Third generation sequencing